spectec/spec/7-module.watsup

```diff
@@ -0,0 +1,236 @@
+  -- if ma = |$meminst((s; f))|
+  -- if s_new = s[.MEM =.. mi]
+  -- if (s_res, ma'*) = $alloc_mem((s_new; f), mem'*)
+
+
+def $alloc_elem(state, elem*) : (store, elemaddr*)
+
+def $alloc_elem((s; f), epsilon) = (s, epsilon)
+
+def $alloc_elem((s; f), elem elem'*) = (s_res, ea ea'*)
+  -- if elem = ELEM reftype (instr*)* elemmode?
+  -- (Step_read : s; f; instr* ~> ref)*
+  -- if ei = { TYPE reftype, ELEM ref* }
+  -- if ea = |$eleminst((s; f))|
+  -- if s_new = s[.ELEM =.. ei]
+  -- if (s_res, ea'*) = $alloc_elem((s_new; f), elem'*)
+
+
+def $alloc_data(state, data*) : (store, dataaddr*)
+
+def $alloc_data((s; f), epsilon) = (s, epsilon)
+
+def $alloc_data((s; f), data data'*) = (s_res, da da'*)
+  -- if data = DATA byte* datamode?
+  -- if di = { DATA byte* }
+  -- if da = |$datainst((s; f))|
+  -- if s_new = s[.DATA =.. di]
+  -- if (s_res, da'*) = $alloc_data((s_new; f), data'*)
+
+
+def $replace_moduleinst(store, funcaddr*, moduleinst) : store
+
+def $replace_moduleinst(s, epsilon, m) = s
+
+def $replace_moduleinst(s, fa fa'*, m) = s_res
+  -- if s_new = s[.FUNC[fa].MODULE = m]
+  -- if s_res = $replace_moduleinst(s_new, fa'*, m)
+
+
+def $alloc_export(moduleinst, export) : exportinst
+
+def $alloc_export(m, export) = xi
+  -- if export = EXPORT name externuse
+  -- if externuse = FUNC x
+  -- if externval = FUNC m.FUNC[x]
+  -- if xi = { NAME name, VALUE externval }
+
+def $alloc_export(m, export) = xi
+  -- if export = EXPORT name externuse
+  -- if externuse = GLOBAL x
+  -- if externval = GLOBAL m.GLOBAL[x]
+  -- if xi = { NAME name, VALUE externval }
+
+def $alloc_export(m, export) = xi
+  -- if export = EXPORT name externuse
+  -- if externuse = TABLE x
+  -- if externval = TABLE m.TABLE[x]
+  -- if xi = { NAME name, VALUE externval }
+
+def $alloc_export(m, export) = xi
+  -- if export = EXPORT name externuse
+  -- if externuse = MEM x
+  -- if externval = MEM m.MEM[x]
+  -- if xi = { NAME name, VALUE externval }
+
+
+def $alloc_module(store, module, externval*) : (store, moduleinst)
+
+def $alloc_module(s, module, externval*) = (s_res, m_res)
+  -- if module = MODULE import* func* global* table* mem* elem* data* start? export*
+  -- if m_init = { FUNC epsilon, GLOBAL epsilon, TABLE epsilon, MEM epsilon, ELEM epsilon, DATA epsilon, EXPORT epsilon }
+  -- if m_im = $alloc_import(m_init, import*, externval*)
+  -- if f = { LOCAL epsilon, MODULE m_im }
+  -- if (s_func, fa*) = $alloc_func((s; f), func*)
+  -- if (s_global, ga*) = $alloc_global((s_func; f), global*)
+  -- if (s_table, ta*) = $alloc_table((s_global; f), table*)
+  -- if (s_mem, ma*) = $alloc_mem((s_table; f), mem*)
+  -- if (s_elem, ea*) = $alloc_elem((s_mem; f), elem*)
+  -- if (s_data, da*) = $alloc_data((s_elem; f), data*)
+  -- if m_ex = m_im ++ { FUNC fa*, GLOBAL ga*, TABLE ta*, MEM ma*, ELEM ea*, DATA da*, EXPORT epsilon }
+  -- if xi* = $alloc_export(m_ex, export)*
+  -- if m_res = m_ex[.EXPORT = xi*]
+  -- if s_res = $replace_moduleinst(s_data, fa*, m_res)
```

Comment:
I see why you did it this way, but now the alloc_module functions is not actually allocating one module instance, but an arbitrary number of them. Can we change this so that there is only one module instance being created in the end?

Also, Conrad suggested we can avoid resorting to mutating immutable parts of the store and instead break the cycle by precomputing the function addresses and the module instance before creating the function instances.